torchdp/privacy_engine.py

```diff
@@ -95,16 +116,33 @@ def get_privacy_spent(self, target_delta: float = None):
 
     def step(self):
         self.steps += 1
-        clip_values = self.clipper.step()
+        clip_values, batch_size = self.clipper.step()
+
+        # ensure the clipper consumed the right amount of gradients.
+        # In the last batch of a training epoch, we might get a batch that is
+        # smaller than others but we should never get a batch that is too large
+        if batch_size > self.batch_size:
+            raise ValueError(
```

Comment:
I like this ValueError, but this contradicts what you mentioned earlier in the comments on imagenet.py. I left some comments there. I would may be also add a Warning for batch_size < self.batch_size which just prints out about the potential non-optimal accumulation in the last rounds of the loop.

Reply:
This ValueError is not raised in the imagenet.py example because we're calling `optimizer.zero_grad()` at the beginning of each epoch. So any accumulated, but not yet finalized gradients get erased.
I'll add the warning.

Reply:
Cool